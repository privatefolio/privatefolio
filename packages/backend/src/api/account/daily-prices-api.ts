import { PAIR_MAPPER, PRICE_APIS, PRICE_MAPPER } from "src/extensions/prices/providers"
import {
  ChartData,
  DailyPrice,
  MyAsset,
  ProgressCallback,
  ResolutionString,
  SqlParam,
  SubscriptionChannel,
  TaskPriority,
  TaskTrigger,
  Time,
  Timestamp,
} from "src/interfaces"
import { allPriceApiIds } from "src/settings/price-apis"
import { PRICE_API_PAGINATION, PRICE_APIS_META } from "src/settings/settings"
import { getAssetTicker } from "src/utils/assets-utils"
import { formatDate } from "src/utils/formatting-utils"
import { createSubscription } from "src/utils/sub-utils"
import { isTestEnvironment, noop } from "src/utils/utils"

import { getAccount } from "../accounts-api"
import { getMyAssets, patchAsset } from "./assets-api"
import { enqueueTask } from "./server-tasks-api"

type NewDailyPrice = Omit<DailyPrice, "id">

function deriveDailyPriceId(record: NewDailyPrice) {
  return `${record.assetId}_${record.timestamp}`
}

export async function upsertDailyPrices(
  accountName: string,
  records: DailyPrice[] | NewDailyPrice[]
) {
  const account = await getAccount(accountName)

  try {
    await account.executeMany(
      `INSERT OR REPLACE INTO daily_prices (
        id, assetId, timestamp, price, pair, priceApiId
      ) VALUES (?, ?, ?, ?, ?, ?)`,
      records.map((record) => [
        "id" in record ? record.id : deriveDailyPriceId(record),
        record.assetId,
        record.timestamp,
        JSON.stringify(record.price) || null,
        record.pair || null,
        record.priceApiId || null,
      ])
    )
    account.eventEmitter.emit(SubscriptionChannel.DailyPrices)
  } catch (error) {
    throw new Error(`Failed to add or replace daily prices: ${error}`)
  }
}

export async function upsertDailyPrice(accountName: string, record: DailyPrice | NewDailyPrice) {
  return upsertDailyPrices(accountName, [record])
}

export async function getPricesForAsset(
  accountName: string,
  assetId: string,
  timestamp?: Timestamp,
  start?: Timestamp,
  end?: Timestamp
): Promise<ChartData[]> {
  if (assetId === "USDT" && !!timestamp) {
    return [{ time: timestamp / 1000, value: 1 }] as ChartData[]
  }

  const account = await getAccount(accountName)

  let query = `
    SELECT price FROM daily_prices
    WHERE assetId = ? 
  `
  if (timestamp && (start || end)) {
    throw new Error("You can only pass timestamp or start/end, not both")
  }

  const params: SqlParam[] = []

  if (timestamp) {
    query += " AND timestamp = ?"
    params.push(timestamp)
  }
  if (start && end) {
    query += " AND timestamp >= ? AND timestamp <= ?"
    params.push(start, end)
  } else if (start) {
    query += " AND timestamp >= ?"
    params.push(start)
  } else if (end) {
    query += " AND timestamp <= ?"
    params.push(end)
  }

  query += " ORDER BY timestamp ASC"

  const prices = await account.execute(query, [assetId, ...params])

  return prices.map((x) => JSON.parse(x[0] as string))
}

export async function getAssetPriceMap(
  accountName: string,
  timestamp: Timestamp = new Date().getTime()
): Promise<Record<string, ChartData>> {
  const account = await getAccount(accountName)

  const day: Timestamp = timestamp - (timestamp % 86400000)

  const prices = await account.execute(
    `
    SELECT assetId, price FROM daily_prices
    WHERE timestamp = ?
  `,
    [day]
  )

  return prices.reduce(
    (map: Record<string, ChartData>, x) => {
      map[x[0] as string] = JSON.parse(x[1] as string)
      return map
    },
    {
      USDT: { time: (day / 1000) as Time, value: 1 },
    }
  )
}

export async function getPriceCursor(
  accountName: string,
  assetId: string
): Promise<Timestamp | undefined> {
  const account = await getAccount(accountName)

  const query = `
    SELECT timestamp FROM daily_prices
    WHERE assetId = ?
    ORDER BY timestamp DESC
    LIMIT 1
  `
  const prices = await account.execute(query, [assetId])

  if (prices.length === 0) {
    return undefined
  }

  return JSON.parse(prices[0][0] as string)
}

type FetchDailyPricesOptions = {
  /**
   * less than or equal to this timestamp
   */
  until?: number
}

export async function fetchDailyPrices(
  accountName: string,
  assetsParam?: MyAsset[],
  progress: ProgressCallback = noop,
  signal?: AbortSignal,
  options: FetchDailyPricesOptions = {}
) {
  let assets: MyAsset[]

  if (!assetsParam) {
    assets = await getMyAssets(accountName)
  } else {
    assets = assetsParam
  }

  await progress([0, `Fetching asset prices for ${assets.length} assets`])

  const now = Date.now()
  const today: Timestamp = now - (now % 86400000)

  const promises: (() => Promise<void>)[] = []

  for (let i = 1; i <= assets.length; i++) {
    // eslint-disable-next-line no-loop-func
    promises.push(async () => {
      const asset = assets[i - 1]

      if (!asset.coingeckoId && !isTestEnvironment) {
        await progress([undefined, `Skipped ${getAssetTicker(asset.id)}: No coingeckoId`])
        return
      }

      if (signal?.aborted) throw new Error(signal.reason)

      const preferredPriceApiId = asset.priceApiId
      const priceApiIds = preferredPriceApiId ? [preferredPriceApiId] : allPriceApiIds

      let since: Timestamp | undefined = await getPriceCursor(accountName, asset.id)
      let until: Timestamp | undefined = options.until || today

      if (!since) since = until - 86400000 * PRICE_API_PAGINATION - 1
      if (since === until) since = since - 1

      for (const priceApiId of priceApiIds) {
        try {
          while (true) {
            const priceApi = PRICE_APIS[priceApiId]
            const priceMapper = PRICE_MAPPER[priceApiId]
            const pairMapper = PAIR_MAPPER[priceApiId]

            if (!priceApi || !priceMapper || !pairMapper) {
              throw new Error(`Price API "${priceApiId}" is not supported`)
            }

            const pair = pairMapper(asset.id)

            const request = {
              limit: PRICE_API_PAGINATION,
              pair,
              since,
              timeInterval: "1d" as ResolutionString,
              until,
            }

            // console.log("Daily price api request:", request)

            const results = await priceApi(request)

            if (!preferredPriceApiId && results.length > 0) {
              patchAsset(accountName, asset.id, { priceApiId })
            }

            if (signal?.aborted) throw new Error(signal.reason)

            if (results.length === 0) {
              throw new Error(`${PRICE_APIS_META[priceApiId].name}: EmptyResponse`)
            }

            const documents = results.map((result) => {
              const price = priceMapper(result)
              const timestamp = (price.time as number) * 1000

              const doc: NewDailyPrice = {
                assetId: asset.id,
                pair,
                price,
                priceApiId,
                timestamp,
              }

              return doc
            })

            await upsertDailyPrices(accountName, documents)

            const start = (priceMapper(results[0]).time as number) * 1000
            const end = (priceMapper(results[results.length - 1]).time as number) * 1000

            await progress([
              undefined,
              `Fetched ${getAssetTicker(asset.id)} using ${
                PRICE_APIS_META[priceApiId].name
              } from ${formatDate(start)} to ${formatDate(end)}`,
            ])

            if (results.length !== PRICE_API_PAGINATION) {
              // reached listing date (genesis)
              break
            }

            until = start - 86400000
            since = start - 86400000 * PRICE_API_PAGINATION
          }
          break
        } catch (error) {
          await progress([undefined, `Skipped ${getAssetTicker(asset.id)}: ${error}`])
        }
      }
    })
  }

  let progressCount = 0

  await Promise.all(
    promises.map((fetchFn) =>
      fetchFn().then(async () => {
        progressCount += 1
        await progress([(progressCount / assets.length) * 100])
      })
    )
  )
}

export function enqueueFetchPrices(accountName: string, trigger: TaskTrigger) {
  return enqueueTask(accountName, {
    description: "Fetching price data for all assets.",
    determinate: true,
    function: async (progress, signal) => {
      await fetchDailyPrices(accountName, undefined, progress, signal)
    },
    name: "Fetch asset prices",
    priority: TaskPriority.Low,
    trigger,
  })
}

export async function deleteDailyPrices(accountName: string) {
  const account = await getAccount(accountName)
  await account.execute("DELETE FROM daily_prices")
}

export function enqueueDeleteAssetPrices(accountName: string, trigger: TaskTrigger) {
  return enqueueTask(accountName, {
    description: "Deleting price data for all assets.",
    determinate: true,
    function: async () => {
      await deleteDailyPrices(accountName)
    },
    name: "Delete asset prices",
    priority: TaskPriority.High,
    trigger,
  })
}

export async function subscribeToDailyPrices(accountName: string, callback: () => void) {
  return createSubscription(accountName, SubscriptionChannel.DailyPrices, callback)
}
