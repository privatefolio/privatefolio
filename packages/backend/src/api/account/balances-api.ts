import Big from "big.js"
import { getErc20Balance, getNativeBalance } from "src/extensions/utils/evm-utils"
import { EventCause, ResolutionString, SqlParam } from "src/interfaces"
import { PLATFORMS_META } from "src/settings/platforms"
import { getAssetContract, getAssetPlatform, isEvmPlatform } from "src/utils/assets-utils"
import { formatDate, ONE_DAY } from "src/utils/formatting-utils"
import { floorTimestamp, noop } from "src/utils/utils"

import {
  Balance,
  BalanceMap,
  ProgressCallback,
  SubscriptionChannel,
  TaskPriority,
  TaskTrigger,
  Timestamp,
} from "../../interfaces"
import { DB_OPERATION_PAGE_SIZE } from "../../settings/settings"
import { getAccount } from "../accounts-api"
import { getMyAssets } from "./assets-api"
import { countAuditLogs, getAuditLogOrderQuery, getAuditLogs, getWallets } from "./audit-logs-api"
import { getPricesForAsset } from "./daily-prices-api"
import { getValue, setValue } from "./kv-api"
import { getPlatform } from "./platforms-api"
import { enqueueTask } from "./server-tasks-api"

export async function getBalances(
  accountName: string,
  query = "SELECT * FROM balances ORDER BY timestamp ASC",
  params?: SqlParam[]
): Promise<BalanceMap[]> {
  const account = await getAccount(accountName)

  try {
    const result = await account.execute(query, params)
    return result.map((row) => {
      const value: BalanceMap = JSON.parse(row[1] as string)
      value.timestamp = row[0] as Timestamp
      return value
    })
  } catch (error) {
    throw new Error(`Failed to query balances: ${error}`)
  }
}

export async function invalidateBalances(accountName: string, newValue: Timestamp) {
  const existing = (await getValue<Timestamp>(accountName, "balancesCursor", 0)) as Timestamp

  if (newValue < existing) {
    await setValue(accountName, "balancesCursor", newValue)
  }
}

export async function getBalancesAt(
  accountName: string,
  cursor: Timestamp = -1
): Promise<Balance[]> {
  const account = await getAccount(accountName)
  let balancesCursor =
    cursor !== -1
      ? cursor
      : ((await getValue<Timestamp>(accountName, "balancesCursor", 0)) as Timestamp)

  balancesCursor = floorTimestamp(balancesCursor, "1D" as ResolutionString)

  try {
    const result = await account.execute("SELECT * FROM balances WHERE timestamp = ?", [
      balancesCursor,
    ])
    const [timestamp, data] = result[0]
    const map = JSON.parse(data as string) as BalanceMap
    const balanceDocs = Object.keys(map).map((x) => ({ assetId: x, balance: map[x] }))

    const balances = await Promise.all(
      balanceDocs.map(async (x) => {
        const prices = await getPricesForAsset(accountName, x.assetId, timestamp as number)
        const price = prices.length > 0 ? prices[0] : undefined

        return {
          assetId: x.assetId,
          balance: x.balance as string,
          balanceN: Number(x.balance),
          id: `${timestamp as number}_${x.assetId}`,
          price,
          value: price ? price.value * Number(x.balance) : undefined,
        }
      })
    )

    return balances
  } catch (error) {
    return []
  }
}

export type GetHistoricalBalancesRequest = {
  limit?: number
  order?: "asc" | "desc"
  skip?: number
  symbol?: string
}

export type ComputeBalancesRequest = {
  pageSize?: number
  since?: Timestamp
  until?: Timestamp
}

export async function computeBalances(
  accountName: string,
  request: ComputeBalancesRequest = {},
  progress: ProgressCallback = noop,
  signal?: AbortSignal
) {
  const { pageSize = DB_OPERATION_PAGE_SIZE, until = Date.now() } = request
  let { since } = request

  // TODO8 skip assets tagged as spam

  const account = await getAccount(accountName)
  if (since === undefined) {
    since = (await getValue<Timestamp>(accountName, "balancesCursor", 0)) as Timestamp
  }

  if (since !== 0) {
    await progress([0, `Refreshing balances starting ${formatDate(since)}`])
  }

  const count =
    since === 0
      ? await countAuditLogs(accountName)
      : await countAuditLogs(accountName, "SELECT COUNT(*) FROM audit_logs WHERE timestamp >= ?", [
          since,
        ])
  await progress([0, `Computing balances for ${count} audit logs`])

  const orderQueryAsc = await getAuditLogOrderQuery(true)
  const orderQueryDesc = await getAuditLogOrderQuery()

  let recordsLength = 0
  let latestBalances: Omit<BalanceMap, "timestamp"> = {}
  const walletBalances: Record<string, Record<string, string>> = {} // wallet -> assetId -> balance

  let genesisDay: Timestamp = -1

  if (since !== 0) {
    try {
      const result = await account.execute("SELECT data FROM balances WHERE timestamp = ?", [
        since - 86400000,
      ])
      const data = result[0][0]
      latestBalances = typeof data === "string" ? JSON.parse(data) : latestBalances

      // initialize walletBalances; going through each balance and each wallet
      const wallets = await getWallets(accountName)
      for (const balance of Object.keys(latestBalances)) {
        for (const wallet of wallets) {
          const [latestLog] = await getAuditLogs(
            accountName,
            `SELECT * FROM audit_logs WHERE wallet = ? AND assetId = ? AND timestamp <= ? ${orderQueryDesc} LIMIT 1`,
            [wallet, balance, since]
          )
          if (!walletBalances[wallet]) {
            walletBalances[wallet] = {}
          }
          walletBalances[wallet][balance] = latestLog?.balanceWallet || "0"
        }
      }
    } catch {
      // ignore
    }
  } else {
    genesisDay = 0
  }

  let historicalBalances: Record<number, Omit<BalanceMap, "timestamp">> = {}

  let latestDay: Timestamp = 0

  for (let i = 0; i < count; i += pageSize) {
    if (signal?.aborted) throw new Error(signal.reason)

    const firstIndex = i + 1
    const lastIndex = Math.min(i + pageSize, count)

    await progress([Math.floor((i * 90) / count), `Processing logs ${firstIndex} to ${lastIndex}`])
    const logs = await getAuditLogs(
      accountName,
      `SELECT * FROM audit_logs WHERE timestamp >= ? ${orderQueryAsc} LIMIT ? OFFSET ?`,
      [since, pageSize, i]
    )

    for (const log of logs) {
      const { assetId, change, timestamp, wallet } = log

      if (genesisDay === 0) {
        genesisDay = floorTimestamp(timestamp, "1D" as ResolutionString)
      }

      const nextDay: Timestamp = floorTimestamp(timestamp, "1D" as ResolutionString)

      // fill the daily gaps
      if (latestDay !== 0) {
        const daysDiff = (nextDay - latestDay) / 86400000
        if (daysDiff > 1) {
          for (let i = 1; i < daysDiff; i++) {
            const gapDay = latestDay + i * 86400000
            historicalBalances[gapDay] = Object.assign({}, latestBalances)
          }
        }
      }

      // Initialize wallet balances if needed
      if (!walletBalances[wallet]) {
        walletBalances[wallet] = {}
      }

      // Update per-wallet balance
      if (!walletBalances[wallet][assetId]) {
        walletBalances[wallet][assetId] = change
      } else {
        walletBalances[wallet][assetId] = new Big(walletBalances[wallet][assetId])
          .plus(new Big(change))
          .toFixed()
      }

      // Update cumulative balance (for historical balances table)
      if (!latestBalances[assetId]) {
        latestBalances[assetId] = change
      } else {
        latestBalances[assetId] = new Big(latestBalances[assetId]).plus(new Big(change)).toFixed()
      }

      // Update audit log
      log.balance = latestBalances[assetId] as string
      log.balanceWallet = walletBalances[wallet][assetId] as string

      // Remove zero wallet balances
      if (walletBalances[wallet][assetId] === "0") {
        delete walletBalances[wallet][assetId]
      }

      // remove zero balances from cumulative
      if (latestBalances[assetId] === "0") {
        delete latestBalances[assetId]
      }

      // update historical balances
      if (!historicalBalances[nextDay]) {
        historicalBalances[nextDay] = Object.assign({}, latestBalances)
      } else if (latestBalances[assetId] !== "0") {
        historicalBalances[nextDay][assetId] = latestBalances[assetId]
      }

      latestDay = nextDay
    }

    await account.executeMany(
      "UPDATE audit_logs SET balance = ?, balanceWallet = ? WHERE id = ?",
      logs.map((log) => [log.balance as string, log.balanceWallet as string, log.id])
    )
    account.eventEmitter.emit(SubscriptionChannel.AuditLogs, EventCause.Updated)

    const balanceIds = Object.keys(historicalBalances)

    recordsLength += balanceIds.length
    await account.executeMany(
      "INSERT OR REPLACE INTO balances (timestamp, data) VALUES (?, ?)",
      // eslint-disable-next-line no-loop-func
      balanceIds.map((timestamp) => [timestamp, JSON.stringify(historicalBalances[timestamp])])
    )

    await progress([
      Math.floor((Math.min(i + pageSize, count) * 90) / count),
      `Processed ${balanceIds.length} daily balances`,
    ])

    // free memory
    historicalBalances = {}
  }

  if (latestDay === 0 && since === 0) {
    await account.execute("DELETE FROM balances")
    return
  }
  if (latestDay === 0) latestDay = since

  await progress([96, `Filling balances to reach today`])
  for (let i = latestDay + 86400000; i <= until; i += 86400000) {
    historicalBalances[i] = latestBalances
    latestDay = i
  }

  // filling the day before the genesis day
  if (genesisDay > 0) {
    historicalBalances[genesisDay - ONE_DAY] = {}
  }

  if (Object.keys(historicalBalances).length > 0) {
    const balanceIds = Object.keys(historicalBalances)
    await account.executeMany(
      "INSERT OR REPLACE INTO balances (timestamp, data) VALUES (?, ?)",
      balanceIds.map((timestamp) => [
        Number(timestamp),
        JSON.stringify(historicalBalances[timestamp]),
      ])
    )
    recordsLength += balanceIds.length
  }

  await progress([99, `Setting balances cursor to ${formatDate(latestDay)}`])
  await setValue(accountName, "balancesCursor", latestDay)
  await progress([100, `Saved ${recordsLength} records to disk`])
}

export function enqueueRecomputeBalances(accountName: string, trigger: TaskTrigger) {
  return enqueueTask(accountName, {
    description: "Recomputing balances of owned assets.",
    determinate: true,
    function: async (progress, signal) => {
      await computeBalances(accountName, { since: 0 }, progress, signal)
    },
    name: "Recompute balances",
    priority: TaskPriority.Medium,
    trigger,
  })
}

export function enqueueRefreshBalances(accountName: string, trigger: TaskTrigger) {
  return enqueueTask(accountName, {
    description: "Refreshing balances of owned assets.",
    determinate: true,
    function: async (progress, signal) => {
      await computeBalances(accountName, undefined, progress, signal)
    },
    name: "Refresh balances",
    priority: TaskPriority.Medium,
    trigger,
  })
}

export async function deleteBalances(accountName: string) {
  const account = await getAccount(accountName)
  await account.execute("DELETE FROM balances")
  await account.execute("UPDATE audit_logs SET balance=null, balanceWallet=null")
  account.eventEmitter.emit(SubscriptionChannel.AuditLogs, EventCause.Updated)
  account.eventEmitter.emit(SubscriptionChannel.Balances)
}

export function enqueueDeleteBalances(accountName: string, trigger: TaskTrigger) {
  return enqueueTask(accountName, {
    description: "Deleting all balances.",
    determinate: true,
    function: async () => {
      await deleteBalances(accountName)
    },
    name: "Delete balances",
    priority: TaskPriority.Low,
    trigger,
  })
}

/**
 * Checks on-chain balances for all wallets/assets and compares to stored balances.
 * Returns a list of discrepancies: { wallet, assetId, onChain, stored }
 */
export async function verifyBalances(
  accountName: string,
  progress: ProgressCallback = noop,
  signal?: AbortSignal
) {
  await progress([0, "Loading wallets and assets"])
  const wallets = await getWallets(accountName)
  const assets = await getMyAssets(accountName)

  if (signal?.aborted) throw new Error(signal.reason)

  // Filter to only EVM assets with coingeckoId
  const evmAssets = assets.filter((asset) => {
    if (!asset.coingeckoId) return false
    const platform = getAssetPlatform(asset.id)
    return platform && isEvmPlatform(platform) && PLATFORMS_META[platform]?.chainId
  })

  const totalChecks = wallets.length * evmAssets.length
  await progress([5, `Checking ${totalChecks} wallet/asset combinations`])

  const discrepancies: Array<{
    assetId: string
    onChain: string
    platform: string
    stored: string | undefined
    wallet: string
  }> = []

  let currentCheck = 0
  const orderQuery = await getAuditLogOrderQuery()

  for (const wallet of wallets) {
    for (const asset of evmAssets) {
      if (signal?.aborted) throw new Error(signal.reason)

      currentCheck++
      const progressPercent = 5 + Math.floor((currentCheck * 85) / totalChecks)

      if (currentCheck % 10 === 0) {
        await progress([
          progressPercent,
          // `Checking ${asset.symbol || asset.id} balance for wallet ${wallet} (${currentCheck}/${totalChecks})`,
        ])
      }

      const assetId = asset.id
      const platformId = getAssetPlatform(assetId)
      const platform = await getPlatform(platformId)
      const meta = PLATFORMS_META[platformId]
      const contract = getAssetContract(assetId)
      let onChain: string | undefined

      try {
        if (contract === "0x0000000000000000000000000000000000000000") {
          // Native asset
          onChain = await getNativeBalance(wallet, meta.chainId)
        } else {
          // ERC-20
          const { balance } = await getErc20Balance(contract, wallet, meta.chainId)
          onChain = balance
        }
      } catch (error) {
        await progress([
          progressPercent,
          `Error fetching ${asset.symbol || asset.id} for ${wallet}: ${error instanceof Error ? error.message : error}`,
        ])
        continue
      }

      // Get the latest balance for this wallet/asset combination
      const [latestLog] = await getAuditLogs(
        accountName,
        `SELECT * FROM audit_logs WHERE wallet = ? AND assetId = ? ${orderQuery} LIMIT 1`,
        [wallet, assetId]
      )
      if (!latestLog) continue
      if (latestLog && !latestLog.balanceWallet) {
        throw new Error(`No balance found on audit log. Did you forget to call "computeBalances"?`)
      }
      const stored = latestLog.balanceWallet

      const onChainBig = onChain ? new Big(onChain) : new Big(0)
      const storedBig = stored ? new Big(stored) : new Big(0)

      const difference = onChainBig.minus(storedBig)

      if (difference.gt(0) || difference.lt(0)) {
        const onChainStr = onChainBig.toFixed()
        const storedStr = storedBig.toFixed()
        const diffStr = difference.toFixed()

        const gravity = difference.lt(0) ? "Error" : "Warning"

        await progress([
          progressPercent,
          `${gravity}: ${platform.name} wallet ${wallet} has a mismatch of ${asset.symbol || asset.id}: expected ${storedStr} but got ${onChainStr} (diff: ${diffStr})`,
        ])

        discrepancies.push({
          assetId,
          onChain: onChainStr,
          platform: platformId,
          stored: storedStr,
          wallet,
        })
      }
    }
  }

  await progress([undefined, `Found ${discrepancies.length} balance discrepancies`])
  await progress([100, `Verification complete - ${discrepancies.length} discrepancies found`])
  return discrepancies
}

export function enqueueVerifyBalances(accountName: string, trigger: TaskTrigger) {
  return enqueueTask(accountName, {
    description: "Verifying balances of owned assets.",
    determinate: true,
    function: async (progress, signal) => {
      await verifyBalances(accountName, progress, signal)
    },
    name: "Verify balances",
    priority: TaskPriority.Low,
    trigger,
  })
}
