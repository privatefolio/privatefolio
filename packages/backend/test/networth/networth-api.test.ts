import fs from "fs"
import { join } from "path"
import { computeBalances } from "src/api/account/balances-api"
import { fetchDailyPrices } from "src/api/account/daily-prices-api"
import { importFile } from "src/api/account/file-imports-api"
import { computeNetworth, getNetworth } from "src/api/account/networth-api"
import { ProgressUpdate, ResolutionString } from "src/interfaces"
import { aggregateCandles } from "src/utils/data-utils"
import { beforeAll, expect, it, vi } from "vitest"

const accountName = Math.random().toString(36).substring(7)

const mocks = vi.hoisted(() => {
  return {
    access: vi.fn().mockResolvedValue(true),
    readFile: vi.fn(),
  }
})

beforeAll(async () => {
  vi.mock("fs/promises", () => ({
    ...fs.promises,
    access: mocks.access,
    readFile: mocks.readFile,
  }))
  //
  const fileName = "coinmama.csv"
  const filePath = join("test/files", fileName)
  const buffer = await fs.promises.readFile(filePath, "utf8")
  mocks.readFile.mockResolvedValue(buffer)
  await importFile(accountName, {
    createdBy: "user",
    id: 0,
    metadata: {
      lastModified: 0,
      size: buffer.length,
      type: "text/csv",
    },
    name: fileName,
    scheduledAt: 0,
    status: "completed",
  })
  const until = Date.UTC(2017, 8, 14, 0, 0, 0, 0) // 14 Sep 2017
  await computeBalances(accountName, { until })
  await fetchDailyPrices(accountName, [
    {
      coingeckoId: "bitcoin",
      id: "coinmama:BTC",
      priceApiId: "coinbase",
      symbol: "BTC",
    },
  ])
})

it.sequential("should compute historical networth", async () => {
  // arrange
  // act
  const updates: ProgressUpdate[] = []
  await computeNetworth(accountName, 0, async (state) => updates.push(state))
  const networthArray = await getNetworth(accountName)
  // assert
  expect(updates.join("\n")).toMatchInlineSnapshot(
    `
    "5,Computing networth for 15 days
    10,Computing networth starting Aug 31, 2017
    95,Saving 15 records to the database
    99,Setting networth cursor to Sep 14, 2017"
  `
  )
  expect(networthArray.length).toMatchInlineSnapshot(`15`)
  expect(networthArray).toMatchSnapshot()
  expect(aggregateCandles(networthArray, "3d" as ResolutionString)).toMatchSnapshot("as 3d candles")
  expect(aggregateCandles(networthArray, "1w" as ResolutionString)).toMatchSnapshot("as 1w candles")
  expect(aggregateCandles(networthArray, "1m" as ResolutionString)).toMatchSnapshot("as 1m candles")
})

it.sequential("should refresh networth", async () => {
  // arrange
  // act
  const updates: ProgressUpdate[] = []
  await computeNetworth(accountName, undefined, async (state) => updates.push(state))
  const networthArray = await getNetworth(accountName)
  // assert
  expect(updates.join("\n")).toMatchInlineSnapshot(`
    "5,Computing networth for 1 days
    10,Computing networth starting Sep 14, 2017
    95,Saving 1 records to the database
    99,Setting networth cursor to Sep 14, 2017"
  `)
  expect(networthArray.length).toMatchInlineSnapshot(`15`)
  expect(networthArray).toMatchSnapshot()
})
