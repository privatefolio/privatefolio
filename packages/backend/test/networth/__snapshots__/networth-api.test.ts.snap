// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should compute historical networth > as 1m candles 1`] = `
[
  {
    "close": 0,
    "high": 0,
    "low": 0,
    "open": 0,
    "time": 1501545600,
    "value": 0,
  },
  {
    "close": 237.14,
    "high": 337.37,
    "low": 159.48,
    "open": 0,
    "time": 1504224000,
    "value": 237.14,
  },
]
`;

exports[`should compute historical networth > as 1w candles 1`] = `
[
  {
    "close": 166.44,
    "high": 178.03,
    "low": 0,
    "open": 0,
    "time": 1503878400,
    "value": 166.44,
  },
  {
    "close": 310.17,
    "high": 337.37,
    "low": 159.48,
    "open": 166.44,
    "time": 1504483200,
    "value": 310.17,
  },
  {
    "close": 237.14,
    "high": 307.21,
    "low": 237.14,
    "open": 310.17,
    "time": 1505088000,
    "value": 237.14,
  },
]
`;

exports[`should compute historical networth > as 3d candles 1`] = `
[
  {
    "close": 167.3,
    "high": 178.03,
    "low": 0,
    "open": 0,
    "time": 1504137600,
    "value": 167.3,
  },
  {
    "close": 159.48,
    "high": 166.44,
    "low": 159.48,
    "open": 167.3,
    "time": 1504396800,
    "value": 159.48,
  },
  {
    "close": 317.37,
    "high": 337.37,
    "low": 166.09,
    "open": 159.48,
    "time": 1504656000,
    "value": 317.37,
  },
  {
    "close": 307.21,
    "high": 316.23,
    "low": 307.21,
    "open": 317.37,
    "time": 1504915200,
    "value": 307.21,
  },
  {
    "close": 237.14,
    "high": 303.84,
    "low": 237.14,
    "open": 307.21,
    "time": 1505174400,
    "value": 237.14,
  },
]
`;

exports[`should compute historical networth 3`] = `
[
  {
    "change": 0,
    "changePercentage": 0,
    "time": 1504137600,
    "timestamp": 1504137600000,
    "value": 0,
  },
  {
    "change": 178.03,
    "changePercentage": 0,
    "time": 1504224000,
    "timestamp": 1504224000000,
    "value": 178.03,
  },
  {
    "change": -10.73,
    "changePercentage": -6.03,
    "time": 1504310400,
    "timestamp": 1504310400000,
    "value": 167.3,
  },
  {
    "change": -0.86,
    "changePercentage": -0.51,
    "time": 1504396800,
    "timestamp": 1504396800000,
    "value": 166.44,
  },
  {
    "change": -4.6,
    "changePercentage": -2.76,
    "time": 1504483200,
    "timestamp": 1504483200000,
    "value": 161.84,
  },
  {
    "change": -2.36,
    "changePercentage": -1.46,
    "time": 1504569600,
    "timestamp": 1504569600000,
    "value": 159.48,
  },
  {
    "change": 6.61,
    "changePercentage": 4.14,
    "time": 1504656000,
    "timestamp": 1504656000000,
    "value": 166.09,
  },
  {
    "change": 171.28,
    "changePercentage": 103.12,
    "time": 1504742400,
    "timestamp": 1504742400000,
    "value": 337.37,
  },
  {
    "change": -20,
    "changePercentage": -5.93,
    "time": 1504828800,
    "timestamp": 1504828800000,
    "value": 317.37,
  },
  {
    "change": -1.14,
    "changePercentage": -0.36,
    "time": 1504915200,
    "timestamp": 1504915200000,
    "value": 316.23,
  },
  {
    "change": -6.06,
    "changePercentage": -1.92,
    "time": 1505001600,
    "timestamp": 1505001600000,
    "value": 310.17,
  },
  {
    "change": -2.96,
    "changePercentage": -0.95,
    "time": 1505088000,
    "timestamp": 1505088000000,
    "value": 307.21,
  },
  {
    "change": -3.37,
    "changePercentage": -1.1,
    "time": 1505174400,
    "timestamp": 1505174400000,
    "value": 303.84,
  },
  {
    "change": -22.56,
    "changePercentage": -7.42,
    "time": 1505260800,
    "timestamp": 1505260800000,
    "value": 281.28,
  },
  {
    "change": -44.14,
    "changePercentage": -15.69,
    "time": 1505347200,
    "timestamp": 1505347200000,
    "value": 237.14,
  },
]
`;

exports[`should refresh networth 3`] = `
[
  {
    "change": 0,
    "changePercentage": 0,
    "time": 1504137600,
    "timestamp": 1504137600000,
    "value": 0,
  },
  {
    "change": 178.03,
    "changePercentage": 0,
    "time": 1504224000,
    "timestamp": 1504224000000,
    "value": 178.03,
  },
  {
    "change": -10.73,
    "changePercentage": -6.03,
    "time": 1504310400,
    "timestamp": 1504310400000,
    "value": 167.3,
  },
  {
    "change": -0.86,
    "changePercentage": -0.51,
    "time": 1504396800,
    "timestamp": 1504396800000,
    "value": 166.44,
  },
  {
    "change": -4.6,
    "changePercentage": -2.76,
    "time": 1504483200,
    "timestamp": 1504483200000,
    "value": 161.84,
  },
  {
    "change": -2.36,
    "changePercentage": -1.46,
    "time": 1504569600,
    "timestamp": 1504569600000,
    "value": 159.48,
  },
  {
    "change": 6.61,
    "changePercentage": 4.14,
    "time": 1504656000,
    "timestamp": 1504656000000,
    "value": 166.09,
  },
  {
    "change": 171.28,
    "changePercentage": 103.12,
    "time": 1504742400,
    "timestamp": 1504742400000,
    "value": 337.37,
  },
  {
    "change": -20,
    "changePercentage": -5.93,
    "time": 1504828800,
    "timestamp": 1504828800000,
    "value": 317.37,
  },
  {
    "change": -1.14,
    "changePercentage": -0.36,
    "time": 1504915200,
    "timestamp": 1504915200000,
    "value": 316.23,
  },
  {
    "change": -6.06,
    "changePercentage": -1.92,
    "time": 1505001600,
    "timestamp": 1505001600000,
    "value": 310.17,
  },
  {
    "change": -2.96,
    "changePercentage": -0.95,
    "time": 1505088000,
    "timestamp": 1505088000000,
    "value": 307.21,
  },
  {
    "change": -3.37,
    "changePercentage": -1.1,
    "time": 1505174400,
    "timestamp": 1505174400000,
    "value": 303.84,
  },
  {
    "change": -22.56,
    "changePercentage": -7.42,
    "time": 1505260800,
    "timestamp": 1505260800000,
    "value": 281.28,
  },
  {
    "change": 0,
    "changePercentage": 0,
    "time": 1505347200,
    "timestamp": 1505347200000,
    "value": 237.14,
  },
]
`;
